booster/library/Booster/Pattern/Binary.hs

```diff
@@ -106,8 +138,40 @@ insertInternedString pos str =
         lift $
             modify (\s@DecoderState{internedStrings} -> s{internedStrings = Map.insert pos str internedStrings})
 
-areCompatible :: Version -> Version -> Bool
-areCompatible a b = a.major == b.major && a.minor == b.minor
+-- | Insert a new item into the term store or return a previously-seen instance
+registerTerm :: ShallowTerm -> DecodeM (Idx, Term)
+registerTerm shallow = DecodeM $ do
+    ds@DecoderState{termCache, termStore} <- lift get
+    case Map.lookup shallow termCache of
+        Just idx -> pure (idx, termStore `Seq.index` idx)
+        Nothing -> do
+            let !new = resolve termStore shallow -- strict: fail early on inconsistent data
+                newStore = termStore :|> new
+                newIdx = Seq.length termStore -- NB index 0-based
+                newCache = Map.insert shallow newIdx termCache
+            lift $ put ds{termStore = newStore, termCache = newCache}
+            pure (newIdx, new)
+
+getTerm :: Idx -> DecodeM Term
+getTerm idx = DecodeM $ do
+    store <- lift $ gets termStore
+    pure $ store `Seq.index` idx
+
+{- | Resolves indexes into the term store in a shallow term, assuming
+all indexes exist. Returns the resolved (full) term.
+-}
+resolve :: Seq Term -> ShallowTerm -> Term
+resolve store (ShallowTerm shallow) = case shallow of
+    AndTermF i1 i2 -> AndTerm (fromStore i1) (fromStore i2)
+    SymbolApplicationF sym sorts is -> SymbolApplication sym sorts $ map fromStore is
+    DomainValueF sort payload -> DomainValue sort payload
+    VarF v -> Var v
+    InjectionF s1 s2 i -> Injection s1 s2 (fromStore i)
+    other -> error $ "Unexpected shallow term " <> show other
```

Comment:
I'm feeling a little uneasy about this call to `error`, but I do agree that it will not ever be triggered by construction of the shallow terms in `resolve -> mkSymbolApplication`.